core/src/debuginfo/source_manager.rs

```diff
@@ -0,0 +1,441 @@
+    }
+    #[inline(always)]
+    fn get(&self, id: SourceId) -> Result<Arc<SourceFile>, SourceManagerError> {
+        (**self).get(id)
+    }
+    #[inline(always)]
+    fn get_by_path(&self, path: &str) -> Option<Arc<SourceFile>> {
+        (**self).get_by_path(path)
+    }
+    #[inline(always)]
+    fn find(&self, name: &str) -> Option<SourceId> {
+        (**self).find(name)
+    }
+    #[inline(always)]
+    fn file_line_col_to_span(&self, loc: FileLineCol) -> Option<SourceSpan> {
+        (**self).file_line_col_to_span(loc)
+    }
+    #[inline(always)]
+    fn file_line_col(&self, span: SourceSpan) -> Result<FileLineCol, SourceManagerError> {
+        (**self).file_line_col(span)
+    }
+    #[inline(always)]
+    fn location_to_span(&self, loc: Location) -> Option<SourceSpan> {
+        (**self).location_to_span(loc)
+    }
+    #[inline(always)]
+    fn location(&self, span: SourceSpan) -> Result<Location, SourceManagerError> {
+        (**self).location(span)
+    }
+    #[inline(always)]
+    fn source(&self, id: SourceId) -> Result<&str, SourceManagerError> {
+        (**self).source(id)
+    }
+    #[inline(always)]
+    fn source_slice(&self, span: SourceSpan) -> Result<&str, SourceManagerError> {
+        (**self).source_slice(span)
+    }
+}
+
+#[cfg(feature = "std")]
+pub trait SourceManagerExt: SourceManager {
+    /// Load the content of `path` into this [SourceManager], using the given path as the source
+    /// name.
+    fn load_file(&self, path: &std::path::Path) -> Result<Arc<SourceFile>, SourceManagerError> {
+        let name = path.to_string_lossy();
+        if let Some(existing) = self.get_by_path(name.as_ref()) {
+            return Ok(existing);
+        }
+
+        let name = Arc::from(name.into_owned().into_boxed_str());
+        let content = std::fs::read_to_string(path)
+            .map(|s| SourceContent::new(Arc::clone(&name), s.into_boxed_str()))
+            .map_err(SourceManagerError::LoadFailed)?;
+
+        Ok(self.load_from_raw_parts(name, content))
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T: ?Sized + SourceManager> SourceManagerExt for T {}
+
+// DEFAULT SOURCE MANAGER
+// ================================================================================================
+
+use crate::utils::sync::RwLock;
+
+#[derive(Default)]
+pub struct MultiThreadedSourceManager(RwLock<DefaultSourceManagerImpl>);
+impl Clone for MultiThreadedSourceManager {
+    fn clone(&self) -> Self {
+        let manager = self.0.read();
+        Self(RwLock::new(manager.clone()))
+    }
+}
+
+#[derive(Default)]
+pub struct SingleThreadedSourceManager(RefCell<DefaultSourceManagerImpl>);
+impl Clone for SingleThreadedSourceManager {
+    fn clone(&self) -> Self {
+        let manager = self.0.borrow();
+        Self(RefCell::new(manager.clone()))
+    }
+}
+
+#[derive(Default, Clone)]
+struct DefaultSourceManagerImpl {
+    files: Vec<Arc<SourceFile>>,
+    names: BTreeMap<Arc<str>, SourceId>,
+}
```

Comment:
A couple of questions about this setup:

First, It seems like we need `RwLock` only for the `MultiThreadedSourceManager`. Then why do we also need to provide a `no_std` implementation of it? Are there scenarios where a multi-threaded source manager going to be used in `no_std` context? If not, then we can probably just enable `MultiThreadedSourceManager` when `std` feature is enabled.

Next, it seems like `DefaultSourceManagerImpl` is basically the same as `SingleThreadedSourceManager`. Could we define `DefaultSourceManager` as something like this:
```Rust
pub struct DefaultSourceManager {
    files: RefCell<Vec<Arc<SourceFile>>>,
    names: RefCell<BTreeMap<Arc<str>, SourceId>>,
}
```
And then get rid of `SingleThreadedSourceManager`?

Lastly, I'm wondering if the `MultiThreadedSourceManager` should be defined here vs. the compiler. It seems like it is needed exclusively for the compiler so maybe `SourceManager` trait and `DefaultSourceManager` struct stay here, but `MultiThreadedSourceManager` is implemented in the compiler?

Reply:
The reason for a no-std compatible `RwLock` is that it makes it possible to define a `SourceManager` impl in a `static` global variable, a common way of instantiating a source manager for use across the duration of a programs lifetime. I would expect this to be a common downstream use case, and one we ourselves may wish to use. More generally however, I think it would be exceedingly useful to have a synchronization primitive that works on all our supported platforms, and `miden-core` is the place to put it IMO.

Secondly, it allows me to provide default implementations that have "canonical" behavior that are usable in all scenarios, and on all supported platforms. It further allows the implementation details to be shared for these default implementations, rather than duplicated.

`DefaultSourceManagerImpl` provides the common implementation behavior for both single- and multi-threaded source managers, and exports a mutable API so that we aren't doing something bonkers like wrapping a `RefCell` access with a `RwLock`. Both `SourceManager` impls are simple wrappers around it, but that's deliberate.

While the compiler is an obvious user of `MultiThreadedSourceManager`, it is unlikely to be the only one, but even if it is, being able to re-use the default implementation in both single and multi-threaded contexts is worth the tradeoff IMO. It would be really unfortunate if every time we needed a `SourceManager` in a multi-threaded context, we had to write a duplicate of the implementation here.

I'd also be very concerned with naming an implementation with internal `RefCell` fields, `DefaultSourceManager`, because users might then do something dumb like wrap it in an `Arc<Mutex<T>>` or whatever. Being explicit about how it is intended to be used protects against that.

Reply:
> More generally however, I think it would be exceedingly useful to have a synchronization primitive that works on all our supported platforms, and `miden-core` is the place to put it IMO.

I am fine with this - but if we do add such a low level primitive here, we should make sure it is well-tested and I don't see issues with including `loom` as a dev-dependency. It does feel like this is something that could be done in a separate PR though.

I'm also curious if this could help us down the road to build a no-std compatible version of something like [pollster](https://github.com/zesterer/pollster) (this is something we could really benefit from in `miden-base`).

> The reason for a no-std compatible `RwLock` is that it makes it possible to define a `SourceManager` impl in a `static` global variable, a common way of instantiating a source manager for use across the duration of a programs lifetime. I would expect this to be a common downstream use case, and one we ourselves may wish to use.

There is something that doesn't click here for me yet. Specifically: are we intending to use `MultiThreadedSourceManager` in a no-std context? My assumption is that no-std usually means that we are running in a single-threaded environment (e.g., WASM) and the combination of no-std + multi-threaded environment is not something that we need to support.

If the assumption is correct, if we do keep the `MultiThreadedSourceManager` I would enable it only when `std` feature is enabled (and this would also imply that we don't immediately need no-std `RwLock` and it could be added in a separate PR).

More generally though, I'm trying to keep things as simple as possible. For example, if we could make the source manager completely internal to the assembler (i.e., the assembler works only with a default implementation of the source manager), that would simplify things quite a bit. My understanding is that we need to make it more flexible because of how the compiler works. Though, it is not clear to me in what context would the assembler and the complier need to share the same source manager. Basically, if the assembler is single-use (and single-threaded), why not make the source manager for it single use as well? It is possible I'm not seeing far enough here though.

If we don't need to share source managers between the compiler and the assembler, I would rather make assembler opinionated in that it always uses a `SingleThreadedSourceManager`. We could still keep the `MultiThreadedSourceManager` defined here to encapsulate the logic and this way, the compiler could use a `MultiThreadedSourceManager` and the assembler could use a `SingleThreadedSourceManager`.

If we do need to share the source managers between the two, in my mind there are two scenarios we need to support:

1. Compilation on a regular machine - i.e., with `std` and multi-threading enabled.
2. Compilation in WASM - i.e., `no-std` and single-threaded.

If these are sufficient for our purposes, then as I mentioned above `SingleThreadedSourceManager` would always be exported, but `MultiThreadedSourceManager` would be exported only under `std` feature.

Reply:
> I'm also curious if this could help us down the road to build a no-std compatible version of something like [pollster](https://github.com/zesterer/pollster) (this is something we could really benefit from in miden-base).

Having some synchronization primitives that work in no-std is definitely critical to being able to use something like that. I seem to recall there being some no-std compatible async runtimes, but don't have any specifically in mind. Some runtimes provide their own locking primitives, but those are usually bulkier runtimes.

> There is something that doesn't click here for me yet. Specifically: are we intending to use MultiThreadedSourceManager in a no-std context? My assumption is that no-std usually means that we are running in a single-threaded environment (e.g., WASM) and the combination of no-std + multi-threaded environment is not something that we need to support.

The point of having a no-std compatible synchronization primitive, is that we can use `MultiThreadedSourceManager` in our code, and it works when compiling to both std and no-std environments, with no changes. It's no different than using a `Mutex` when your program only has one thread - in such cases, you never contend the lock, so its basically free.

The presumption is that you can default to `MultiThreadedSourceManager` anywhere that you need a `Send + Sync` source manager, and not have to worry about what to do when compiling for Wasm. I used `SingleThreadedSourceManager` in many areas of the code base, for the simple reason that all of those uses were _definitely_ single-threaded uses, so I could explicitly choose the more performant option.

> If the assumption is correct, if we do keep the MultiThreadedSourceManager I would enable it only when std feature is enabled (and this would also imply that we don't immediately need no-std RwLock and it could be added in a separate PR).

For the reason I stated above, that defeats the point of having it at all. It needs to be available in both std and no-std environments, so that you can be largely target-agnostic in your code.

Perhaps the naming here is confusing, maybe it should be `DefaultSourceManager` and `DefaultSourceManagerSync` or something like that. I think I just made it confusing by bringing threading into this, since it isn't so much to do with threads directly, but rather the need for one or the other depending on if your code requires `Send` and `Sync` traits to be implemented on it.

> More generally though, I'm trying to keep things as simple as possible. For example, if we could make the source manager completely internal to the assembler (i.e., the assembler works only with a default implementation of the source manager), that would simplify things quite a bit. My understanding is that we need to make it more flexible because of how the compiler works. Though, it is not clear to me in what context would the assembler and the complier need to share the same source manager. Basically, if the assembler is single-use (and single-threaded), why not make the source manager for it single use as well? It is possible I'm not seeing far enough here though.

The assembler is effectively part of the compiler - it doesn't seem that way because it currently lives in this repo, but in practice, it is important for them to share some of this key infrastructure. Without this PR, many of the types implemented here that were previously defined in `miden-assembly` (e.g. `Span`, `SourceSpan`, etc.) are slightly-incompatible duplicates of similar structures in the compiler, for essentially no good reason other than they started as separate projects. This adds a bunch of friction in the backend of the compiler, as we must translate source information available to the compiler, in to a form the assembler recognizes. A part of this PR is also refactoring how we use those types in the assembler to allow for more precise source-level debugging information to be passed through. By providing a standard set of source management/debug info primitives in `miden-core`, our entire toolchain can share this information seamlessly. I can instantiate a source manager in the compiler's `main` function, and use it to compile from Rust to Miden IR, then to MASM, then assemble that to MAST, then execute that MAST and render assertions with source spans in a diagnostic. That's enabled by having these tools. It makes no sense to me to duplicate these things in each project, when they are so key to compilation and assembly, _and_ lack any real differentiation that would call for having multiple implementations.

I understand, and totally agree with, the goal of keeping things simple, but as the project grows and requires more robust developer tooling, some of this complexity is simply inescapable, and I'd like to avoid having to keep refactoring things because we aren't anticipating our needs well enough. I knew we were going to need this stuff back when I refactored the assembler's parser, but I avoided making the changes then to keep that set of changes simpler - but we ended up needing the changes before we've even gotten to the point where we're using the refactored assembler in our other projects. I do want to emphasize that I'm not adding things to the crates in this repo lightly - I pretty much prefer to keep _everything_ in the compiler repo unless I have no choice. If I'm adding something here, it's because I feel it is clearly beneficial to incur the significant overhead in maintaining those features here, rather than in the compiler - either because I _can't_ do so in the compiler, or because there is a clear need for them in the assembler (or other key crates) as well.

> If we don't need to share source managers between the compiler and the assembler, I would rather make assembler opinionated in that it always uses a SingleThreadedSourceManager. We could still keep the MultiThreadedSourceManager defined here to encapsulate the logic and this way, the compiler could use a MultiThreadedSourceManager and the assembler could use a SingleThreadedSourceManager.

We do need to share them, which is a major motivation for implementing this stuff here. I'm fine with `Assembler::default()` constructing with a `SingleThreadedSourceManager` (already the case IIRC), but `Assembler::new()` must accept an `Arc<dyn SourceManager>` so that other implementations can be used with it.

Reply:
I should also note, that part of the problem I'm attempting to anticpate/solve here, is the fact that the compiler's current source management is not compatible with Wasm. If we want to be able to build the compiler for use in the playground, I need to start making changes to allow that to happen. My intent (and I've already done this in my development branch) is to replace the compiler's source management with this system, and unify both the compiler and assembler in one go. This not only enables source-level debug info to be passed through seamlessly, but it will remove a major component of the compiler that was preventing us from targeting Wasm. 

The `RwLock` impl here will be used for other compiler components that need to be `Send + Sync` (e.g. global symbol table, and other similar structures). While that part could be done in the compiler, it feels like a waste to implement a primitive like `RwLock` in a place where no other project in our toolchain can make use of it, should they need to.

Reply:
Thank you for the detailed explanations - I think this clarifies things quite a bit. So, my current understanding is that `MultiThreadedSourceManager` is basically what is going to be used by the compiler both in multi-threaded contexts and in single-threaded contexts like WASM.

Then, the only that I'm not seeing a benefit of having a `SingleThreadedSourceManager` manager at all. If the multi-threaded version works in either context, why not rename `MultiThreadedSourceManager` into something like `DefaultSourceManager` and use it everywhere? I do understand that there is probably some performance benefit in having `SingleThreadedSourceManager` not rely on `RwLock` at all - but it also sees like `SingleThreadedSourceManager` manger is not going to be used in any performance critical contexts (and if such context does arise, someone could write a non-default implementation that avoids the `RwLock`).

So, it seems to me, it may make sense to provide a single `DefaultSourceManager` which is functionally equivalent to the current `MultiThreadedSourceManager`. Unless I'm missing something?